### ISSUE TITLE
Look into ROC curve for setting thresholds for testing assertions

### ISSUE THREAD

**Issue by aidanjgriffiths** [OnTrack-UG-Squad/speaker-verification#18]

**Comment by aidanjgriffiths** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [Look into ROC curve for setting thresholds for testing assertions](https://trello.com/c/XM1XmBLJ/124-look-into-roc-curve-for-setting-thresholds-for-testing-assertions)

**Comment by MoeHdaib** [comment]
### **ROC curve** 
**Iteration 0 progress**

ROC curve is good way to visualize the performance of our model mainly because its different distinction thresholds. AUC is the area below the ROC curve which conclude the model performance and it ranges from 0-1 as 1 is the best.
In python it we can do it using Scikit-learn machine learning library. Basic example: 

```
# xtest : data samples, mfcc_001, mfcc_002, ...
pred = model.predict(xtest)

# ytest : the expected output -> pass or fail, pred : the predictions -> predict % similarity
#fpr: False positive rate, tpr = True positive rate
fpr, tpr, thresholds = roc_curve(ytest, pred)

auc_k = auc(fpr, tpr)

plt.figure()
plt.plot([0, 1], [0, 1], 'k--')
plt.plot(fpr, tpr, label='Keras (area = {:.3f})'.format(auc_k))
plt.xlabel('False positive rate')
plt.ylabel('True positive rate')
plt.title('ROC curve')
plt.legend(loc='lower right')
plt.show()
```

In order to implement it to the speaker-verification project, we need the data samples similar to those used to train the model with and its respective labels. In our case, it is pass or fail as in if the user is actually what they claim to be. Therefore, we can download and use data samples from LibriSpeech as this was used to train the original model (up for discussion and improvement).

Current contraints: 
- The original [model](https://github.com/philipperemy/deep-speaker/) is trained with categorical crossentropy but our classification is binary. (maybe because different thersholds??)
- If it is worth following the original [model](https://github.com/philipperemy/deep-speaker/) way to evaluate the model as they calculate the ROC but there is many unuseful parts reletive to our project. (to be tested)

**Comment by aidanjgriffiths** [comment]
Closing due to investigation being completed and documented by @MoeHdaib.
